api/v1alpha1/volumereplication_types.go

```diff
@@ -20,20 +20,77 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// ImageState represents the replication operations to be performed on the image
+type ImageState string
+
+const (
+	// Primary ImageState enables mirroring and promotes the image to primary
+	Primary ImageState = "primary"
+
+	// Secondary ImageState demotes the image to secondary and resyncs the image if out of sync
+	Secondary ImageState = "secondary"
+
+	// Resync option resyncs the image
+	Resync ImageState = "resync"
+)
+
+// ReplicationState captures the latest state of the replication operation
+type ReplicationState string
+
+const (
+	// MirroringEnabled means mirroring is enabled between two sites
+	MirroringEnabled ReplicationState = "MirroringEnabled"
+
+	// ImagePromoted means image was promoted to primary
+	ImagePromoted ReplicationState = "Promoted"
+
+	// ImageDemoted means image was demoted to secondary
+	ImageDemoted ReplicationState = "Demoted"
+
+	// ImageResynced means image was resynced
+	ImageResynced ReplicationState = "Resynced"
+
+	// MirroringFailed means mirroring operation was failed
+	MirroringFailed ReplicationState = "MirroringFailed"
```

Comment:
Does this mean enabling mirroring failed?

Reply:
Yes, that was the original intent.  But now we are planning to keep it simple with following options only. 

```
// ReplicationState captures the latest state of the replication operation
type ReplicationState string

const (
	//Success signifies that the last replication operation succeeded
	Success ReplicationState = "Success"

	//Failure signifies that the last replication operation failed
	Failure ReplicationState = "Failure"
)
```
In case of failure, the details will be captured in the `Message` field if the `Status`. 

What do you think?

Reply:
I think it's hard to have a "success" state for something (the replication) that runs continuously. For me, "success" is more a one-time action, which once completed can be reflected as a success.

I can see a state such as "Mirroring" or "Replicating".

The failure makes sense if for whatever reason the replication is stopped and data is not replicated anymore.

Reply:
Is there any intention to add status.conditions? Checking to see if we are elaborating any further conditions in here.

Also, A user can flip the `imageState` from primary(op-1)->secondary(op-2)->primary(op-3) even before the reconciler gets a chance to act on (or complete) the intermediate change (op-2). Thus a status that reflects primary as the result from op-1 spec would be stale as observed by the user. Are there plans to address this? (I would suggest adding to status an observed generation number, like [replicasets](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#replicasetstatus-v1-apps), but there maybe better patterns).

Reply:
> Is there any intention to add status.conditions? Checking to see if we are elaborating any further conditions in here.
>
Yes. That's what I am proposing as a follow-up PR since there are some implementation details to be figured out.
>
> Also, A user can flip the `imageState` from primary(op-1)->secondary(op-2)->primary(op-3) even before the reconciler gets a chance to act on (or complete) the intermediate change (op-2). Thus a status that reflects primary as the result from op-1 spec would be stale as observed by the user. Are there plans to address this? (I would suggest adding to status an observed generation number, like [replicasets](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#replicasetstatus-v1-apps), but there maybe better patterns).
>
But how would observed generation help?
I wouldn't expect the users to check `Status.ObservedGeneration` before making the changes
so eventually when they do decide to check, they wouldn't know what the initial generation was and what should be the latest generation.
Or am I completely missing the point?

Reply:
> > Also, A user can flip the `imageState` from primary(op-1)->secondary(op-2)->primary(op-3) even before the reconciler gets a chance to act on (or complete) the intermediate change (op-2). Thus a status that reflects primary as the result from op-1 spec would be stale as observed by the user. Are there plans to address this? (I would suggest adding to status an observed generation number, like [replicasets](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#replicasetstatus-v1-apps), but there maybe better patterns).
> 
> But how would observed generation help?
> I wouldn't expect the users to check `Status.ObservedGeneration` before making the changes
> so eventually when they do decide to check, they wouldn't know what the initial generation was and what should be the latest generation.
> Or am I completely missing the point?

The user needs to know that the `observedGeneration` is the latest generation or the same generation as the CR is currently. That would be the only check a user can perform.

IOW, you are right, if `observedGeneration` is older than current generation, there is no way for a user to know what was desired in an older generation, and the user needs to wait till the generation catches up to know desired state was reached.